vortex-datafusion/src/plans.rs

```diff
@@ -0,0 +1,477 @@
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) -> DFResult<SendableRecordBatchStream> {
+        assert_eq!(
+            partition, 0,
+            "single partitioning only supported by TakeOperator"
+        );
+
+        let row_indices_stream = self.input.execute(partition, context)?;
+
+        Ok(Box::pin(TakeRowsStream {
+            row_indices_stream,
+            completed: false,
+            output_projection: self.projection.clone(),
+            output_schema: self.output_schema.clone(),
+            vortex_array: self.table.clone(),
+        }))
+    }
+}
+
+/// Stream of outputs emitted by the [TakeRowsExec] physical operator.
+#[pin_project]
+pub(crate) struct TakeRowsStream<F> {
+    // Stream of row indices arriving from upstream operator.
+    #[pin]
+    row_indices_stream: F,
+
+    completed: bool,
+
+    // Projection based on the schema here
+    output_projection: Vec<usize>,
+    output_schema: SchemaRef,
+
+    // The original Vortex array we're taking from
+    vortex_array: StructArray,
+}
+
+impl<F> Stream for TakeRowsStream<F>
+where
+    F: Stream<Item = DFResult<RecordBatch>>,
+{
+    type Item = DFResult<RecordBatch>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+
+        // If `poll_next` has already fired, return None indicating end of the stream.
+        if *this.completed {
+            return Poll::Ready(None);
+        }
+
+        // Get the indices provided by the upstream operator.
+        let record_batch = match ready!(this.row_indices_stream.poll_next(cx)) {
+            None => {
+                // Row indices stream is complete, we are also complete.
+                // This should never happen right now given we only emit one recordbatch upstream.
+                return Poll::Ready(None);
+            }
+            Some(result) => {
+                *this.completed = true;
+                result?
+            }
+        };
+
+        let row_indices =
+            ArrayData::from_arrow(record_batch.column(0).as_primitive::<UInt64Type>(), false)
+                .into_array();
+
+        // If no columns in the output projection, we send back a RecordBatch with empty schema.
+        // This is common for COUNT queries.
+        if this.output_projection.is_empty() {
+            let opts = RecordBatchOptions::new().with_row_count(Some(row_indices.len()));
+            return Poll::Ready(Some(Ok(RecordBatch::try_new_with_options(
+                Arc::new(Schema::empty()),
+                vec![],
+                &opts,
+            )
+            .unwrap())));
+        }
+
+        // Assemble the output columns using the row indices.
+        // NOTE(aduffy): this re-decodes the fields from the filter schema, which is unnecessary.
+        let mut columns = Vec::new();
+        for (output_idx, src_idx) in this.output_projection.iter().enumerate() {
```

Comment:
we should project the array before and then just do a take here

Reply:
done